textalyze.rb

```diff
@@ -1,66 +1,101 @@
+  sanitize(string)
+  string.split(//)
+end
+
+def sanitize(string)
+  string.downcase!
+  string.gsub!(/\W+/, '') # Remove all non-word chars
+end
 
-  counts # This returns the "counts" hash
+# v1.1
+def frequency(string)
+  hash = item_counts(string_to_array(string)) # returns counts hash
+  hash.each do |key, value|
+    hash[key] = value.to_f/string.size.to_f
+  end
+  hash
 end
 
-# "p" prints something to the screen in a way that's friendlier
-# for debugging purposes than print or puts.
+# Notes on v.1.2
+# remove all non word chars - add to sanitize
+# take hash as input
+# round hash values to 2 places
+# force 0.00 format, figure our alignment
+# format as percent in output
+# scale bars to terminal width
+# get the bars to print out correctly next to the values
 
-p item_counts([1,2,1,2,1]) == {1 => 3, 2 => 2}
-p item_counts(["a","b","a","b","a","ZZZ"]) == {"a" => 3, "b" => 2, "ZZZ" => 1}
-p item_counts([]) == {}
-p item_counts(["hi", "hi", "hi"]) == {"hi" => 3}
-p item_counts([true, nil, "dinosaur"]) == {true => 1, nil => 1, "dinosaur" => 1}
-p item_counts(["a","a","A","A"]) == {"a" => 2, "A" => 2}
+def print_histo(stats)
+  ordered_stats = stats.sort
+  print "Calculating letter frequencies...done"
+  ordered_stats.each do |key, value|
+    percent = (value*100).round(2)
+    puts
+    print "[#{key}] [#{'%5.2f' % percent}%] " # This was trial and error.
+    # I don't have full understanding yet
+    print "="*((scaled_value(value, stats.values)))
+  end
+  puts
+end
 
-# Each of the lines above will print out "true" or "false" and collectively
-# act as a sanity check.  Remember that conceptually "x == y"
-# means "are x and y equal?"
-#
-# That is, when you run the code, if any lines print out "false"
-# then you know something is off in your code.
-#
-# This does *not* mean that your code is perfect if each line
-# prints out "true.""  For example,
-#   1. We might have missed a corner case
-#   2. The code does what it should, but is conceptually confused
-#   3. Something else we haven't though of
-#
-# Remember: Option #3 is *always* possible.
-#
-# Think of these like rumble strips on the side of the road.  They're here
-# to tell you when you're veering off the road, not to guarantee you're
-# driving phenomenally. :)
+def scaled_value(num, range)
+  # I had to look up the equation below - it feels like something
+  # I should remember from high school math. Time for a refresher?
+  # ((to_max - to_min)* (num - from_min)) / ((from_max - from_min) + to_min)
+  to_max = terminal_size # Reasonable screen width?
+  to_min = 0
+  from_max = max(range)
+  from_min = 0
+  ((to_max - to_min) * (num - from_min)) / ((from_max - from_min) + to_min)
+end
+
+def terminal_size
+  # found this at  http://stackoverflow.com/a/2352441
+  dimension = `stty size`.split.map { |x| x.to_i }.reverse
+  dimension[0]-13 # first item in array is the x value of the terminal dimension
```

Comment:
Note that this _doesn't_ solve the problem of calculating the bar width dynamically.  That's a solvable problem, but I'll have to sit down a think for a few minutes about how to do it.  Instead, these are all suggestions to clarify the code and ensure the various parts of your program can operate independently without needing deep knowledge of every other part of your program.